packages/wallet/wdk/src/sequence/handlers/guard.ts

```diff
@@ -30,17 +58,34 @@ export class GuardHandler implements Handler {
       }
     }
 
-    // TODO: check if 2FA is required. If it is, return 'actionable'
-
     return {
       address,
       handler: this,
       status: 'ready',
-      handle: async () => {
-        const signature = await this.guard.signEnvelope(request.envelope)
-        await this.signatures.addSignature(request.id, signature)
-        return true
-      },
+      handle: () =>
+        new Promise(async (resolve, reject) => {
+          try {
+            const signature = await this.guard.signEnvelope(request.envelope)
+            await this.signatures.addSignature(request.id, signature)
+            resolve(true)
+          } catch (e) {
+            if (e instanceof Guard.AuthRequiredError) {
+              const respond = async (code: string) => {
+                try {
+                  const signature = await this.guard.signEnvelope(request.envelope, { id: e.id, code })
+                  await this.signatures.addSignature(request.id, signature)
+                  resolve(true)
+                } catch (e) {
+                  reject(e)
+                }
+              }
+              await onPromptCode(e.id, respond)
+              resolve(true)
```

Comment:
Should we solve right away? I don't remember this perfectly, but I think `handle` should not return early.